csrc/multidevice/pipeline.cpp

```diff
@@ -0,0 +1,307 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/builder.h>
+#include <ir/cloner.h>
+#include <ir/printer.h>
+#include <ir/utils.h>
+#include <iter_visitor.h>
+#include <multidevice/pipeline.h>
+#include <multidevice/pipeline_ir.h>
+
+namespace nvfuser {
+
+int PipelineStageDescriptor::running_unique_id_ = 0;
+
+// Utility class used for Pipeline instantiation
+class PipelineBuilder final {
+ public:
+  PipelineBuilder(Pipeline* pipeline) : pipeline_(pipeline) {
+    validate();
+    fillInfo();
+    buildPipelineVals();
+    buildPipelineStages();
+    buildPipelineCommunications();
+  }
+
+ private:
+  // Points to the pipeline to be constructed
+  Pipeline* pipeline_;
+  // Stores the Stages IR of the pipeline
+  std::vector<PipelineStage*> stages;
+  // stores the Vals that will be I/O of each stage
+  std::unordered_map<const PipelineStageDescriptor*, ValSet> StageInputsDesc;
```

Comment:
```suggestion
  std::unordered_map<const PipelineStageDescriptor*, ValSet> stage_inputs_desc_;
```

Same for the other variables